### ISSUE TITLE
NuGet Restore failing with: 'Invalid static method invocation syntax: "[MSBuild]::VersionGreaterThanOrEquals...'

### ISSUE THREAD

**Issue by maljos88** [NuGet/Home#12851]
### NuGet Product Used

NuGet.exe

### Product Version

6.7.0

### Worked before?

6.6.1

### Impact

I'm unable to use this version

### Repro Steps & Context

Hey all, one of my builds is failing during NuGet restore. The last successful build was using v6.6.1. Thanks in advance

Building solution with:
--
  | SolutionFile = C:\buildkite-builds\...\..\SomeFile.sln
  | DotNetConfiguration = Debug
  | Checking for updates from https://www.nuget.org/api/v2/.
  | Currently running NuGet.exe 4.7.1.
  | Updating NuGet.exe to 6.7.0.
  | Update successful.
  | Installing nuget packages...
  | MSBuild auto-detection: using msbuild version '15.3.409.57025' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\bin'.
  | C:\Users\buildkite-agent...\i4m4kg2z.kqm.nugetrestore.targets(65,17): error MSB4186: Invalid static method invocation syntax: "[MSBuild]::VersionGreaterThanOrEquals($([MSBuild]::ValueOrDefault('$(NETCoreSdkVersion)', '0.0')), '8.0')". Method '[MSBuild]::VersionGreaterThanOrEquals' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)).  [C:\Users\buildkite-agent...\4c0py1qz.fj1.nugetinputs.targets]
  |  
  | NuGet restore failed with exit code 1
  | At C:\opt\someDir\NuGet\NuGet.psm1:53 char:13
  | +             throw "NuGet restore failed with exit code " + $proc.Exit ...
  | +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | + CategoryInfo          : OperationStopped: (NuGet restore failed with exit code 1:String) [], RuntimeException
  | + FullyQualifiedErrorId : NuGet restore failed with exit code 1

**Comment by antibarbie** [comment]
hit by the same problem when building old version of our software....

is provoked by commit : https://github.com/NuGet/NuGet.Client/commit/51fb11cb31c4ff1f87347766f4656261fe839628

That makes the .targets file incompatible with MSbuild 15.x (Visual 2017)

**Comment by antibarbie** [comment]
I had to pin a specific nuget version, to workaround the problem on an old cake script, for an old project, so that it would build with Visual 15.9

`
$NUGET_URL = "https://dist.nuget.org/win-x86-commandline/v6.4.2/nuget.exe"
`

**Comment by maljos88** [comment]
Hi, is there a way to update to a specific version using nuget.exe? I'm currently doing:
nuget.exe update -self
But there doesn't seem to be a way to pass a version. Looks like this issue will not be fixed soon

**Comment by CharlesFarris** [comment]
I originally tried solving this problem by rolling back the version of NuGet to 6.6.1 in my TeamCity builds but a co-worker pointed out a simpler solution.  Set the `NuGetAudit` variable to `true` or `false` and the evaluation will get by-passed.  Reference: https://learn.microsoft.com/en-us/nuget/concepts/auditing-packages

**Comment by zivkan** [comment]
The good news is that NuGetAudit is enabled by default in NuGet 6.8, so we removed the condition. Therefore, this error should no longer show up once 6.8 goes GA.

NuGet's versioning has always followed Visual Studio's, minus 11.0. So, if you're using MSBuild 15.9, then NuGet 4.9 was released at the same time (and NuGet within VS is that version). Similarly, MSBuild 15.3 and NuGet 4.3. You'll have the fewest problems if you match the NuGet.exe version to the version of MSBuild you're going to build with. Unfortunately this means you also won't get other bug fixes or new features, but that's also true for the old version of MSBuild, compiler, etc.

Personally, I strongly, **STRONGLY** advice do NOT use nuget.exe (or Azure DevOps' `NuGetCommand`) to restore packages. If all your projects use `PackageReference`, so zero `packages.config` files, then use `msbuild -t:restore` instead. If you have even a single packages.config the, then since MSBuild 16.5, [you can use `msbuild -t:restore -p:RestorePackagesConfig=true`](https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-packagereference-and-packagesconfig-projects-with-msbuild). However, if the 23 people who experienced this issue are using VS 2017, then NuGet in that version of MSBuild doesn't have that feature. So you're stuck using nuget.exe, or upgrading to a newer version of Visual Studio.

This also demonstrates a huge problem with nuget.exe. It's not feasible to maintain backwards compatibility with every version of MSBuild in the past, while also being expected to develop new features. At some point we need to use features not available in very old versions of MSBuild. I didn't intentionally break restore when using a new nuget.exe with an old MSBuild (I did test with MSBuild 16.11), but according to [the Visual Studio support lifecycle document](https://learn.microsoft.com/en-us/lifecycle/products/visual-studio-2017), VS went out of mainstream support last year. It's still in extended support, which is a paid service. I know that using newer versions of VS isn't always easy. The UX might change, there might be different bugs, and if your company doesn't have a Visual Studio subscription, just a once-off purchase, then it's also expensive to upgrade. But supporting old versions has a cost, and this is the balance that executives much higher up the chain than me have decided.

**Comment by quasarea** [comment]
> Personally, I strongly, **STRONGLY** advice do NOT use nuget.exe (or Azure DevOps' `NuGetCommand`) to restore packages. If all your projects use `PackageReference`, so zero `packages.config` files, then use `msbuild -t:restore` instead. If you have even a single packages.config the, then since MSBuild 16.5, [you can use `msbuild -t:restore -p:RestorePackagesConfig=true`](https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-packagereference-and-packagesconfig-projects-with-msbuild). However, if the 23 people who experienced this issue are using VS 2017, then NuGet in that version of MSBuild doesn't have that feature. So you're stuck using nuget.exe, or upgrading to a newer version of Visual Studio.

I was under impression that msbuild restore target was deprecated, especially with credential managers plugins and other changes, googled around but cannot find article that I deducted it from so far.

**Comment by zivkan** [comment]
No. With .NET Core, and now  .NET 5+, `dotnet restore` is the msbuild restore target, and is the only "native" way to restore on Linux and Mac (nuget.exe needs Mono)

**Comment by jeffkl** [comment]
Team Triage: We should discuss how we want to handle changes to NuGet.targets that mean we'll require a specific version of MSBuild.
